### PR TITLE
Replace hardcoded Blocks.FARMLAND checks with FarmBlock

### LINKED ISSUE #1195: [1.21] mayPlaceOn is hardcoded to check for vanilla farmland only
**NeoForge Version:** 21.0.40+

After the recent changes to plantable, it´s no longer as straight forward to support the same crop/plant types as the vanilla farmland.
This is due to the hardcoded check for farmland in `mayPlaceOn` for various blocks.

Could be solved with either a instanceof FarmBlock, a new tag check - or potientially the new `neoforge:villager_farmlands` (Suggested by TelepathicGrunt)

Would need to be changed in the following places:
- `CropBlock#mayPlaceOn`
- `BushBlock#mayPlaceOn`
- `StemBlock#mayPlaceOn`
- `AttatchedStemBlock#mayPlaceOn`
- `PitcherCropBlock#mayPlaceOn`

These are the places that checks for specifically Farmland. There is some other instances for plant/plant-like blocks, that also have different hardcoded mayPlaceOn checks - might be worth looking at too?

Issue requested opened by @TelepathicGrunt, after conversation on Discord.

### LINKED ISSUE #1209: PlantType pull #971 appears to have made custom farmland unreliable
With the replacement of the PlantType system, it appears that custom farmland-type blocks can't natively support crops in a reliable way.

With the new system, it appears as though you're expected to use IBlockExtension#canSustainPlant in your farmland
image

However, CropBlock's implementation means that if you return true for this, you bypass the light check (and the super check, but that's mostly irrelevant here)
image

So that would require you implement the light check yourself
image

There's two problems with this:
1. The method is protected, so it's not accessible without an additional AT we shouldn't need to have
2. The method requires a LevelReader, which canSustainPlant does not provide. You _could_ check & cast BlockGetter, however at that point you're basically saying that any time that method is called externally with a different BlockGetter, you can't check light

So ultimately I have two questions:
1. Should custom farmland now require an AT?
2. Should custom farmland now be unable to check base functionality (light) reliably in its own interface

### PR DESCRIPTION
**Pull request by TelepathicGrunt**
Closes https://github.com/neoforged/NeoForge/issues/1195
Closes https://github.com/neoforged/NeoForge/issues/1209 

Vanilla crops have a hardcoded check for Farmland block. Not even instanceof. So to make things easier for people, switching this to an instanceof check seems logical.

Also, now Rabbits will attack tagged modded farmland blocks so players can suffer rabbit raids

CropBlock's hasSufficientLight method is now AT'ed to public so it can be used by modders for various checks and needs.

### CODE REVIEW

**Comment by Soaryn** [comment]
Potentially was considering other things like Chorus fruit, Cocoa beans, etc. a while back. The tag I came up with then considering that system were
-  `valid_soils/crop`
- `valid_soils/cocoa`
- `valid_soils/chorus_fruit`
- `valid_soils/nether_wart` 
- `valid_soils/bamboo` (fortunately, this one is `minecraft:bamboo_plantable_on`
- `valid_soils/sugar_cane`


This was not necessarily an exhaustive search, but I'd be down for alternative names really. The idea was to be a little more scalpel when it comes to growable blocks. See cocoabeans as a prime example of nonsense: requires jungle logs tag. By tagging jungle logs, you tag it with many other side effecting tags. Arguably this could be a completely separate pr as it definitely inflates the scope ab it, but wanted to bring it up.

------
Something to conside: which direction the perspective should be (would like to still group it together) 
`valid_soils/farmland` vs `valid_soils/crop`

**Comment by TelepathicGrunt** [comment]
Update: was talking with another maintainer and mayPlace is called on client. This means we can't really do the tags safely since vanilla clients can connect to server and wouldn't match same behavior. Could cause desync on block placing. Maybe when Neo has a solution for vanilla clients with regards to neo tags, we can revisit this. But for now, doing an instanceof FarmBlock check seems safest for now.

File: patches/net/minecraft/world/level/block/StemBlock.java.patch [hunk 1]
```diff
@@ -20,7 +26,7 @@
                      BlockPos blockpos = p_222540_.relative(direction);
                      BlockState blockstate = p_222539_.getBlockState(blockpos.below());
 -                    if (p_222539_.getBlockState(blockpos).isAir() && (blockstate.is(Blocks.FARMLAND) || blockstate.is(BlockTags.DIRT))) {
-+                    if (p_222539_.isEmptyBlock(blockpos) && (blockstate.is(Blocks.FARMLAND) || blockstate.is(BlockTags.DIRT))) { // TODO 1.20.3 PORTING: reimplement canSustainPlant check
++                    if (p_222539_.isEmptyBlock(blockpos) && (blockstate.getBlock() instanceof net.minecraft.world.level.block.FarmBlock|| blockstate.is(BlockTags.DIRT))) { // TODO 1.20.3 PORTING: reimplement canSustainPlant check
```
Comment (sciwhiz12) [inline patches/net/minecraft/world/level/block/StemBlock.java.patch]: Missed space before the `||`. :)
```suggestion
+                    if (p_222539_.isEmptyBlock(blockpos) && (blockstate.getBlock() instanceof net.minecraft.world.level.block.FarmBlock || blockstate.is(BlockTags.DIRT))) { // TODO 1.20.3 PORTING: reimplement canSustainPlant check
```

Also, is that porting comment still valid?

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]
Small quibble, and also a potential comment cleanup. The PR looks sound (and I can't think of an obvious way to test it in-game).

File: patches/net/minecraft/world/level/block/StemBlock.java.patch (hunk 1 above)
Comment (TelepathicGrunt) [reply patches/net/minecraft/world/level/block/StemBlock.java.patch]: Fixed

**Review by sciwhiz12 — DISMISSED** [review]

**Review by XFactHD — APPROVED** [review]

**Pull request merged**